File: bestia-zoneserver/src/test/java/net/bestia/zoneserver/chat/ChatCommandServiceTest.java
package net.bestia.zoneserver.chat;

import static org.mockito.Mockito.*;

import java.util.Arrays;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.mockito.Mock;

import net.bestia.model.dao.AccountDAO;
import net.bestia.model.domain.Account;

public class ChatCommandServiceTest {

	private static final long ACC_ID = 10;
	private static final String CMD_TXT = "/known la la";

	private ChatCommandService chatService;

	@Mock
	private ChatCommand chatCmd;
	
	@Mock
	private AccountDAO accDao;
	
	@Mock
	private Account acc;

	@Before
	public void setup() {

		when(chatCmd.isCommand(any())).thenReturn(false);
		when(chatCmd.isCommand(CMD_TXT)).thenReturn(true);
		
		when(accDao.findOne(ACC_ID)).thenReturn(acc);

		chatService = new ChatCommandService(Arrays.asList(chatCmd), accDao);
	}

	@Test(expected = NullPointerException.class)
	public void ctor_nullList_throws() {
		new ChatCommandService(null, accDao);
	}
	
	@Test(expected = NullPointerException.class)
	public void ctor_2argNull_throws() {
		new ChatCommandService(Arrays.asList(chatCmd), null);
	}

	@Test
	public void isChatCommand_containedChatPrefix_true() {
		Assert.assertTrue(chatService.isChatCommand("/known test"));
	}

	@Test
	public void isChatCommand_notContainedChatPrefix_false() {
		Assert.assertFalse(chatService.isChatCommand("/unknown test"));
	}

	@Test
	public void executeChatCommand_validTextCommand_chatCommandIsExecuted() {
		chatService.executeChatCommand(ACC_ID, CMD_TXT);

		verify(chatCmd).executeCommand(acc, CMD_TXT);
	}

	@Test
	public void executeChatCommand_invalidTextCommand_noChatCommandIsExecuted() {
		final String CMD_TXT = "/unknown la la";
		chatService.executeChatCommand(ACC_ID, CMD_TXT);

		verify(chatCmd, times(0)).executeCommand(acc, CMD_TXT);
	}
}

File: bestia-zoneserver/src/test/kotlin/net/bestia/zoneserver/chat/ChatCommandServiceTest.kt
package net.bestia.zoneserver.chat

import com.nhaarman.mockitokotlin2.times
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.whenever
import net.bestia.model.account.AccountRepository
import net.bestia.model.account.Account
import net.bestia.model.account.AccountType
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import java.util.*

@ExtendWith(MockitoExtension::class)
class ChatCommandServiceTest {

  private lateinit var chatService: ChatCommandService

  @Mock
  private lateinit var chatCmd: ChatCommand

  @Mock
  private lateinit var accDao: AccountRepository

  @Mock
  private lateinit var acc: Account

  @BeforeEach
  fun setup() {
    chatService = ChatCommandService(Arrays.asList(chatCmd), accDao)
  }

  @Test
  fun isChatCommand_containedChatPrefix_true() {
    assertTrue(chatService.isChatCommand("/known test"))
  }

  @Test
  fun isChatCommand_notContainedChatPrefix_false() {
    assertFalse(chatService.isChatCommand("#unknown test"))
  }

  @Test
  fun executeChatCommand_validTextCommand_chatCommandIsExecuted() {
    whenever(accDao.findById(ACC_ID)).thenReturn(Optional.of(acc))
    whenever(chatCmd.isCommand(CMD_TXT)).thenReturn(true)
    whenever(chatCmd.requiredUserLevel()).thenReturn(AccountType.USER)
    whenever(acc.accountType).thenReturn(AccountType.GM)
    chatService.executeChatCommand(ACC_ID, CMD_TXT)

    verify(chatCmd).executeCommand(acc, CMD_TXT)
  }

  @Test
  fun executeChatCommand_invalidTextCommand_noChatCommandIsExecuted() {
    whenever(accDao.findById(ACC_ID)).thenReturn(Optional.of(acc))
    val CMD_TXT = "/unknown la la"
    chatService.executeChatCommand(ACC_ID, CMD_TXT)

    verify(chatCmd, times(0)).executeCommand(acc, CMD_TXT)
  }

  companion object {
    private const val ACC_ID: Long = 10
    private const val CMD_TXT = "/known la la"
  }
}
